### DISCUSSION TITLE
ESP32 & DRAM usage: using SPRAM ?

### DISCUSSION THREAD

**Discussion by bmedici** [nlohmann/json#4141] [Q&A]
Hi,

On a specific project where I'm using `nlohmann::json` extensively (thanks for this lib), I'm facing some RAM limitations and the firmware does not compile (link, in fact) anymore.
Of course, this is due tu several factors and not only the lib usage. But analysing my firmware.map file, I see many, many lines and a lot of bytes related to my JSON usage. 
eg: `_ZNSt6vectorIN8nlohmann16json_abi_v3_11_210basic_jsonISt3mapS_NSt7__cxx1112basic_stringIcSt11char_traits`

I hav e two questions:
1) do you know about any bad practice to nice things to do tu play with string and `nlohmann::json` in order to optimise memory usage  ?
2) as data structures can become a huge memory pain, how could be make `nlohmann::json` to use the external PSRAM these chips have (4MB are available) ?

Thanks a lot for any pointer or good practice.

Bruno

**Comment by bmedici** [comment]
Still a valid question, any idea or suggestion ?

**Reply by nlohmann** [reply]
You can pass your own `StringType` as template parameter to `basic_json`, see https://json.nlohmann.me/api/basic_json/ and https://json.nlohmann.me/api/basic_json/string_t/

**Reply by bmedici** [reply]
Thank for your reply. However, I'm not only looking to allocate strings elsewhere, but use SPIRAM entirely for a document.
I have to hold a reference document while building a new version on a regular basis, and compare both to transmit only the difference since the reference document. 
Storing the ref doc, which is moderately big, consumes several kB or tens of kB. I'd like to allocate this outside DRAM !
